api/src/main/java/org/projectnessie/tools/catalog/migration/api/CatalogMigrator.java

```diff
@@ -0,0 +1,277 @@
+
+  public static final String FAILED_IDENTIFIERS_FILE = "failed_identifiers.txt";
+  public static final String FAILED_TO_DELETE_AT_SOURCE_FILE = "failed_to_delete_at_source.txt";
+  public static final String DRY_RUN_FILE = "dry_run_identifiers.txt";
+
+  private CatalogMigrator() {}
+
+  /**
+   * Register or Migrate tables from one catalog(source catalog) to another catalog(target catalog).
+   *
+   * <p>Users must make sure that no in-progress commits on the tables of source catalog during
+   * registration.
+   *
+   * @param catalogMigratorParams configuration params
+   * @return List of successfully registered/migrated and list of failed to register/migrate table
+   *     identifiers.
+   */
+  public static CatalogMigrationResult registerTables(CatalogMigratorParams catalogMigratorParams) {
+
+    PrintWriter printWriter = catalogMigratorParams.printWriter();
+    boolean deleteEntriesFromSourceCatalog = catalogMigratorParams.deleteEntriesFromSourceCatalog();
+    String operation = deleteEntriesFromSourceCatalog ? "migration" : "registration";
+
+    List<TableIdentifier> identifiers;
+    if (catalogMigratorParams.tableIdentifiers() == null
+        || catalogMigratorParams.tableIdentifiers().isEmpty()) {
+      identifiers =
+          getMatchingTableIdentifiers(
+              catalogMigratorParams.sourceCatalog(),
+              catalogMigratorParams.identifierRegex(),
+              printWriter);
+    } else {
+      identifiers = catalogMigratorParams.tableIdentifiers();
+    }
+
+    printWriter.println(
+        String.format("%nIdentified %d tables for %s.", identifiers.size(), operation));
+
+    if (catalogMigratorParams.isDryRun()) {
+      CatalogMigrationResult result =
+          ImmutableCatalogMigrationResult.builder()
+              .registeredTableIdentifiers(identifiers)
+              .failedToRegisterTableIdentifiers(Collections.emptyList())
+              .failedToDeleteTableIdentifiers(Collections.emptyList())
+              .build();
+      printWriter.println("Dry run is completed.");
+
+      writeToFile(
+          pathWithOutputDir(catalogMigratorParams.outputDirPath(), DRY_RUN_FILE),
+          result.registeredTableIdentifiers());
+      result.printDryRunResults(printWriter, deleteEntriesFromSourceCatalog);
+      return result;
+    }
+
+    if (deleteEntriesFromSourceCatalog
+        && catalogMigratorParams.sourceCatalog() instanceof HadoopCatalog) {
+      printWriter.println(
+          String.format(
+              "[WARNING]: Source catalog type is HADOOP and it doesn't support dropping tables just from "
+                  + "catalog. %nAvoid operating the migrated tables from the source catalog after migration. "
+                  + "Use the tables from target catalog."));
+    }
+
+    printWriter.println(String.format("%nStarted %s ...", operation));
+    List<TableIdentifier> registeredTableIdentifiers = new ArrayList<>();
```

Comment:
Why all these collections, although ImmutableCatalogMigrationResult.Builder already provides the necessary `add*` methods. So also here - overly complex.